kkrings/chapter03/exercise_3.11.8.cpp

```diff
@@ -0,0 +1,359 @@
+/**
+ * Solutions to exercise 3.11.3, 3.11.7, and 3.11.8
+ */
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+
+/**
+ * @brief Stack implementation for a generic value type
+ *
+ * The stack implementation supports a custom size, where the default
+ * size is 4096, and is based on a non-dynamic array approach.
+ */
+template<typename T, unsigned int max_size = 4096>
+class stack
+{
+    public:
+        stack() : size_(0), data_(new T[max_size]) {};
+
+        stack(const stack& other) : size_(other.size()), data_(new T[max_size])
+        {
+            std::copy(&other.data_[0], &other.data_[other.size()], &data_[0]);
+        }
+
+        stack(stack&&) = default;
+
+        ~stack() = default;
+
+        stack& operator=(const stack& other)
+        {
+            assert(other.size() == size());
+            std::copy(&other.data_[0], &other.data_[other.size()], &data_[0]);
+            return *this;
+        }
+
+        stack& operator=(stack&&) = default;
+
+        /**
+         * @return Show last element.
+         */
+        T& top()
+        {
+            if (empty())
+            {
+                throw std::underflow_error("The stack is empty.");
+            }
+
+            return data_[size_ - 1];
+        }
+
+        /**
+         * Remove last element.
+         */
+        void pop()
+        {
+            if (empty())
+            {
+                throw std::underflow_error("The stack is empty.");
+            }
+
+            --size_;
+        }
+
+        /**
+         * Insert new element.
+         * @param element New element
+         */
+        void push(const T& element)
+        {
+            if (full())
+            {
+                throw std::overflow_error("The stack is full.");
+            }
+
+            data_[size_] = element;
+            ++size_;
+        }
+
+        /**
+         * Delete all entries;
+         */
+        void clear()
+        {
+            size_ = 0;
+        }
+
+        /**
+         * @return Number of elements
+         */
+        unsigned int size() const
+        {
+            return size_;
+        }
+
+        /**
+         * @return Check whether stack is full.
+         */
+        bool full() const
+        {
+            return size() == max_size;
+        }
+
+        /**
+         * @return Check whether stack is empty.
+         */
+        bool empty() const
+        {
+            return size() == 0;
+        }
+
+    private:
+        std::unique_ptr<T[]> data_;
+        unsigned int size_;
```

Comment:
Constructors initialize `size_` first, so it should be the first variable in the class.